JSONB functions and operators

lhirata opened the pull request:
Closes #2731.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e73aa538f58bcb34826eae95669e3fadbef84173/

justinj commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[v2.0/jsonb.md, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7oo1axIJZNMLEePz_s:-L7oo1axIJZNMLEePz_t:bcjmlbf) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L59)):*
> ```Markdown
> Function | Description
> ---------|------------
> `jsonb_build_object(<any_element>...)` | Builds a `JSONB` object out of a variadic argument list.
> ```

if we can spare a bit more verbosity (i know this is supposed to be short) it might be good to mention that this is alternating keys and values.

---

*[v2.0/jsonb.md, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7ooDKrdjqH3t-O4LZX:-L7ooDKrdjqH3t-O4LZY:beb1m1m) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L60)):*
> ```Markdown
> ---------|------------
> `jsonb_build_object(<any_element>...)` | Builds a `JSONB` object out of a variadic argument list.
> `jsonb_each(<jsonb>)` | Expands the outermost `JSONB` object into a set of key value pairs.
> ```

I forgot to mention `json_array_elements` which I think is also very useful!

---

*[v2.0/jsonb.md, line 71 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7onn9jqJp4fAum7MBw:-L7onn9jqJp4fAum7MBx:b-lj0mph) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L71)):*
> ```Markdown
> ---------|-------------|---------|
> `->` | Access a `JSONB` field, returning a JSON value. | `'{"foo":"bar"}'::JSONB->'foo'` = `'bar'::JSONB`
> `-->` | Access a `JSONB` field, returning a string. | `'{"foo":"bar"}'::JSONB->>'foo'` = `'bar'::STRING`
> ```

operator column says `-->` instead of `->>`

---

*[v2.0/jsonb.md, line 72 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7on0VwpPgPeE3FQMr0:-L7on0VwpPgPeE3FQMr1:b-8k7kui) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L72)):*
> ```Markdown
> `->` | Access a `JSONB` field, returning a JSON value. | `'{"foo":"bar"}'::JSONB->'foo'` = `'bar'::JSONB`
> `-->` | Access a `JSONB` field, returning a string. | `'{"foo":"bar"}'::JSONB->>'foo'` = `'bar'::STRING`
> `@>` | Path-value check. | `'{"foo": {"baz": 3}, "bar": 2}'::JSONB @> '{"foo": {"baz": 3}}'::JSONB"` = `true`
> ```

the semantics of this this are weird (and copied from Postgres), I don't know if a full explanation is needed or if we have a good spot for one but [here](https://www.postgresql.org/docs/9.5/static/datatype-json.html#JSON-CONTAINMENT) is Postgres's.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2742)*
<!-- Sent from Reviewable.io -->

lhirata commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[v2.0/jsonb.md, line 59 at r2](https://beta.reviewable.io/reviews/cockroachdb/docs/2742#-L7oo1axIJZNMLEePz_s:-L7vglrH66lB4ph6KPzy:bph8aex) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L59)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

if we can spare a bit more verbosity (i know this is supposed to be short) it might be good to mention that this is alternating keys and values.
</blockquote></details>

added - double checking: the list is what alternates between keys and values, right?

---

*[v2.0/jsonb.md, line 60 at r2](https://beta.reviewable.io/reviews/cockroachdb/docs/2742#-L7ooDKrdjqH3t-O4LZX:-L7vguxx-ZfWOn-3aeXq:bqm0fg3) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L60)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I forgot to mention `json_array_elements` which I think is also very useful!
</blockquote></details>

will add. is there any reason this function (or any of the other functions) is `json_` instead of `jsonb_`? 

---

*[v2.0/jsonb.md, line 71 at r2](https://beta.reviewable.io/reviews/cockroachdb/docs/2742#-L7onn9jqJp4fAum7MBw:-L7vlbTo6ufwqk3eoq4q:b-7ql1td) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L71)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

operator column says `-->` instead of `->>`
</blockquote></details>

fixed

---

*[v2.0/jsonb.md, line 72 at r2](https://beta.reviewable.io/reviews/cockroachdb/docs/2742#-L7on0VwpPgPeE3FQMr0:-L7vo8hz4YRSvdirdK6L:bw7bc8q) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L72)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

the semantics of this this are weird (and copied from Postgres), I don't know if a full explanation is needed or if we have a good spot for one but [here](https://www.postgresql.org/docs/9.5/static/datatype-json.html#JSON-CONTAINMENT) is Postgres's.
</blockquote></details>

took a stab at improving the description

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/docs/2742)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/21b6d8e7af3f6b0511452b90638f60276b424e9c/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/78cccb272a834fec48701542dcfdef020dbd2752/

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[v2.0/jsonb.md, line 59 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7oo1axIJZNMLEePz_s:-L7yhZb2NegX2-oXgLXU:b-1wprt5) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L59)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

added - double checking: the list is what alternates between keys and values, right?
</blockquote></details>

yeah it's like `jsonb_build_object('key1', 'value1', 'key2', 'value2', ...`

---

*[v2.0/jsonb.md, line 60 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-L7ooDKrdjqH3t-O4LZX:-L7yhqQK_E661hmbIavd:b-vykif7) ([raw file](https://github.com/cockroachdb/docs/blob/e73aa538f58bcb34826eae95669e3fadbef84173/v2.0/jsonb.md#L60)):*
<details><summary><i>Previously, lhirata wrote…</i></summary><blockquote>

will add. is there any reason this function (or any of the other functions) is `json_` instead of `jsonb_`? 
</blockquote></details>

no they mostly should have both variations, I guess we should default to `jsonb`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2742#-:-L7yiDltZECQdBC6u8vU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f50e0d1b32849a89637e574e2079d653a9face63/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/ecbf93659d996e49e2d38d175ec445068c2077cf/

jseldess commented on v2.0/jsonb.md:
```diff
@@ -52,6 +52,28 @@ Examples:
+---------|------------
+`jsonb_array_elements(<jsonb>)` | Expands a `JSONB` array to a set of `JSONB` values.
+`jsonb_build_object(<any_element>...)` | Builds a `JSONB` object out of a variadic argument list that alternates between keys and values.
+`jsonb_each(<jsonb>)` | Expands the outermost `JSONB` object into a set of key value pairs.
```
nit: `key-value pairs`

jseldess commented on v2.0/jsonb.md:
```diff
@@ -52,6 +52,28 @@ Examples:
+`jsonb_array_elements(<jsonb>)` | Expands a `JSONB` array to a set of `JSONB` values.
+`jsonb_build_object(<any_element>...)` | Builds a `JSONB` object out of a variadic argument list that alternates between keys and values.
+`jsonb_each(<jsonb>)` | Expands the outermost `JSONB` object into a set of key value pairs.
+`jsonb_object_keys(<jsonb>)` | Returns sorted set of keys in the outermost JSON object.
```
nit: JSON > `JSONB`

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2584ba56ca778776d87435fdd5e2e5eec067a770/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2584ba56ca778776d87435fdd5e2e5eec067a770/
